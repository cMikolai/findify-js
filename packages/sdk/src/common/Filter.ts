/**
 * Filter that can be use to
 * narrow down the search query.
 */
export interface Filter {
  /** Name of filter */
  name: string;
  /** Type of filter */
  type: string;
  /** Array of selected filter values */
  values?: FilterValue[];
}

/*
 * Represents selected filter value.
 */
export interface FilterValue {
  /** Selected filter value */
  value?: string;
  /** Selected range filter "from" value */
  from?: string;
  /** Selected range filter "to" value */
  to?: string;
}
