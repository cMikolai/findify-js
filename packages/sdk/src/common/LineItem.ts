/**
 * Product line item.
 */
export interface LineItem {
  /** Item ID bought */
  item_id: string;
  /** Sale price of the product */
  unit_price: number;
  /** Quantity bought */
  quantity: number;
}
