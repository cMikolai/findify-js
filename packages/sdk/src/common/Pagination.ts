export interface Pagination {
  /**  Offset of products items. Used in pagination */
  offset?: number;
  /** Number of products in result */
  limit?: number;
}
