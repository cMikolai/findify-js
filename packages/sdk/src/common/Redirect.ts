export interface Redirect {
  name: string;
  url: string;
}
