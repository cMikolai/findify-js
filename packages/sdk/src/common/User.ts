/**
 * Defines the current user,
 * which gives just enough information to enable personalization.
 * User object has two properties that are required: `uid` and `sid`.
 */
export interface User {
  /**
   * This is a unique identifier of the user across multiple sessions.
   * You can store the uid in the cookie or get it from any other source (like a database).
   * It must uniquly identify the user.
   * */
  uid: string;
  /**
   * A unique session identifier,
   * the best is to store the value in a cookie on the client side,
   * with a TTL of 30 minutes or use a session ID that your platform provides.
   */
  sid: string;
  /**
   * The email from the user, if the user is logged in.
   * Providing an email allows us to track the user across
   * devices and give them relevant personalized results.
   */
  email?: string;
  /**
   * Not required in browser environment.
   * By default we get the ip from the request that we receive.
   * If call is made not from the user's browser,
   * please include the relevant user ip in the request.
   */
  ip?: string;
  /**
   * Not required in browser environment.
   * By default we get the user agent from the request that we receive.
   * If call is made not from the user's browser,
   * please include the relevant user user agent in the request.
   */
  ua?: string;
  /**
   * Not required in browser environment.
   * By default we get the user languages from the request that we receive.
   * If call is made not from the user's browser,
   * please include the relevant user user languages in the request.
   *
   * You can use either an ISO 639-1 language format code or
   * include the locale code: en-GB.
   */
  lang?: string[];
}
