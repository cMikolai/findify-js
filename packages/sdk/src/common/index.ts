export { Filter } from './Filter';
export { Redirect } from './Redirect';
export { Sort, SortingOrder } from './Sort';
export { User } from './User';
export { Pagination } from './Pagination';
export { LineItem } from './LineItem';
