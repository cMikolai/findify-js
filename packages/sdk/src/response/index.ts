import { Autocomplete } from './Autocomplete';
import { Search } from './Search';
import { SmartCollection } from './SmartCollection';
import { Recommendations } from './Recommendations';
import { Feedback } from './Feedback';

/**
 * Response metadata.
 */
export interface Meta {
  /** Request ID */
  rid: string;
}

/** Response body */
export type Body = CommonBody & SpecificBody;

/**
 * Common response body.
 */
export interface CommonBody {
  meta: Meta;
}

/**
 * Response body depending on the response type.
 */
export type SpecificBody =
  | Autocomplete
  | Search
  | SmartCollection
  | Recommendations
  | Feedback;

export { Autocomplete, Search, SmartCollection, Recommendations, Feedback };
