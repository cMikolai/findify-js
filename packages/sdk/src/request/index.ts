import { User } from '../common';
import * as Autocomplete from './Autocomplete';
import * as Search from './Search';
import * as SmartCollection from './SmartCollection';
import * as Recommendations from './Recommendations';
import * as Feedback from './Feedback';
import * as Content from './Content';
import * as Params from './params';

/**
 * Request method.
 */
export enum Method {
  POST = 'post',
  JSONP = 'jsonp',
}

/** Request body */
export type Body = CommonParams & SpecificParams;

/**
 * Represents a request.
 */
export type Request =
  | Autocomplete.Request
  | Search.Request
  | SmartCollection.Request
  | Recommendations.Request
  | Feedback.Request
  | Content.Request;


/**
 * Common request body parameters.
 */
export interface CommonParams {
  /** A timestamp from the client side of the user */
  t_client: number;
  user: User;
  key: string;
  log?: boolean;
}

/**
 * Request body parameters depending on the request type.
 */
export type SpecificParams =
  | Autocomplete.Params
  | Search.Params
  | SmartCollection.Params
  | Recommendations.Params
  | Feedback.Params;

export { Type } from './Type';
export {
  Params,
  Autocomplete,
  Search,
  SmartCollection,
  Recommendations,
  Feedback,
};
