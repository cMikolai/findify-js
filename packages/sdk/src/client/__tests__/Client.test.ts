import 'dotenv/config';
import debug from 'debug';
import axios from 'axios';
import * as nock from 'nock';
import { Scope as NockScope, NockBackOptions, NockDefinition } from 'nock';
import httpAdapter = require('axios/lib/adapters/http');

import { Client, Config } from '..';
import { User } from '../../common';
import * as Req from '../../request';
import * as Res from '../../response';

import { sanitizeRequestBody, sanitizeResponseBody } from './support';

debug('sdk:tests:client')('TEST_API_HOST: ', process.env.TEST_API_HOST);
debug('sdk:tests:client')('TEST_API_KEY: ', process.env.TEST_API_KEY);
debug('sdk:tests:client')('NOCK_BACK_MODE: ', process.env.NOCK_BACK_MODE);

const user1: User = {
  lang: ['en-GB'],
  uid: 'test1',
  sid: 'ssid1',
  email: 'test2@test.com',
  ip: '192.168.0.1',
  ua:
    'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2228.0 Safari/537.36',
};

const user2: User = {
  lang: ['en-GB'],
  uid: 'test2',
  sid: 'ssid2',
  email: 'test2@test.com',
  ip: '192.168.0.2',
  ua:
    'Mozilla/5.0 (Windows NT 6.2; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/32.0.1667.0 Safari/537.36',
};

function nockBackBeforeHook(scope: NockScope) {
  scope.filteringRequestBody = sanitizeRequestBody as any;
}

const nockBackOpts: NockBackOptions = {
  before: nockBackBeforeHook as any,
  afterRecord: (defs: NockDefinition[]) =>
    defs.map((def: NockDefinition) => ({
      ...def,
      response: sanitizeResponseBody(def.response),
    })),
};

describe('Client', () => {
  beforeAll(async () => {
    // prevents OPTIONS requests, for more info see:
    // https://github.com/node-nock/nock/issues/699#issuecomment-272708264
    // https://github.com/axios/axios/issues/305#issuecomment-272162405
    axios.defaults.adapter = httpAdapter;
    nock.back.fixtures = __dirname + '/__fixtures__';
  });

  afterEach(() => nock.cleanAll());

  const config: Config = {
    url: process.env.TEST_API_HOST,
    key: process.env.TEST_API_KEY,
    method: Req.Method.POST,
    timeout: 5000,
    jsonpCallback: 'findifyCallback',
    retryCount: 1,
    user: user1,
  };

  const client = new Client(config);

  describe('autocomplete', () => {
    it('respects request parameters', async () => {
      const requests: Req.Autocomplete.Request[] = [
        {
          type: Req.Type.Autocomplete,
          params: {
            q: 'tesl',
            item_limit: 2,
            suggestion_limit: 3,
          },
        },
        {
          type: Req.Type.Autocomplete,
          params: {
            q: 'smok',
            item_limit: 3,
            suggestion_limit: 5,
            user: user2,
          },
        },
        {
          type: Req.Type.Autocomplete,
          params: {
            q: 'fir',
            item_limit: 2,
            suggestion_limit: 10,
          },
        },
      ];

      for (let req of requests) {
        const { nockDone, context } = await nock.back(
          'autocomplete.json',
          nockBackOpts
        );
        const response = await client.send(req);
        context.assertScopesFinished();
        expect(sanitizeResponseBody(response)).toMatchSnapshot();
        nockDone();
      }
    });
  });

  it('throws an error on empty query', async () => {
    const queries = ['', '    '];
    const requests: Req.Autocomplete.Request[] = queries.map((q: string) => ({
      type: Req.Type.Autocomplete,
      params: {
        q,
        item_limit: 1,
        suggestion_limit: 1,
      },
    }));

    for (let req of requests) {
      const scope = nock('https://search-staging.findify.io:443')
        .post('/v3/autocomplete', _ => true)
        .reply(400, { error: { message: '"q" is not allowed to be empty' } });
      const response = await client.send(req);
      scope.done();
    }
  });

  describe('search', () => {});

  describe('smart collection', () => {});

  describe('recommendations', () => {});

  describe('feedback', () => {});
});
