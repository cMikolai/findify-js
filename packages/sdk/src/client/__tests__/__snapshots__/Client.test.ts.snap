// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client autocomplete respects request parameters 1`] = `
Object {
  "items": Array [
    Object {
      "availability": true,
      "compare_at": 300,
      "custom_fields": Object {},
      "hash": "AV11VmrACQkVMwqrFH4Y",
      "id": "373056192",
      "price": Array [
        270,
        285,
      ],
      "product_url": "https://smoke-cartel.myshopify.com/products/nexus-glass-tesla-dual-arm-recycler",
      "reviews": Object {
        "average_score": 3.33,
        "total_reviews": 3,
      },
      "stickers": Object {
        "availability": true,
        "discount": Array [
          10,
          5,
        ],
        "free-shipping": false,
        "in-stock": true,
        "out-of-stock": false,
      },
      "thumbnail_url": "https://cdn.shopify.com/s/files/1/0350/8253/products/NEXUS-NEW-TESLA_1_small.jpg?v=1427383897",
      "title": "Nexus Glass \\"Tesla\\" Double Arm Recycler",
    },
  ],
  "meta": Object {
    "item_limit": 2,
    "q": "tesl",
    "rid": "<stripped>",
    "suggestion_limit": 3,
  },
  "suggestions": Array [
    Object {
      "value": "tesla double",
    },
    Object {
      "value": "tesla double arm",
    },
    Object {
      "value": "tesla",
    },
  ],
}
`;

exports[`Client autocomplete respects request parameters 2`] = `
Object {
  "items": Array [
    Object {
      "availability": true,
      "compare_at": -1,
      "custom_fields": Object {},
      "hash": "AV11VmrKCQkVMwqrFIDg",
      "id": "8017104712",
      "price": Array [
        30,
        30,
      ],
      "product_url": "https://smoke-cartel.myshopify.com/products/smoke-cartel-centipede-shenanigans",
      "reviews": Object {
        "average_score": 4.5,
        "total_reviews": 2,
      },
      "stickers": Object {
        "availability": true,
        "discount": false,
        "free-shipping": false,
        "in-stock": true,
        "out-of-stock": false,
      },
      "thumbnail_url": "https://cdn.shopify.com/s/files/1/0350/8253/products/SCHP-SA017-0_small.jpg?v=1476890474",
      "title": "Smoke Cartel \\"Caterpillar\\" Sherlock",
    },
    Object {
      "availability": true,
      "compare_at": -1,
      "custom_fields": Object {},
      "hash": "AV11VniCCQkVMwqrFIHn",
      "id": "9260965640",
      "price": Array [
        29,
        29,
        29,
        29,
        29,
      ],
      "product_url": "https://smoke-cartel.myshopify.com/products/smoke-cartel-crew-shirts",
      "reviews": Object {
        "average_score": 0,
        "total_reviews": 0,
      },
      "stickers": Object {
        "availability": true,
        "discount": false,
        "free-shipping": false,
        "in-stock": true,
        "out-of-stock": false,
      },
      "thumbnail_url": "https://cdn.shopify.com/s/files/1/0350/8253/products/SMOKE-CARTEL-CREW-SHIRTS_SC-T-SHIRTS-WHT-2_small.jpg?v=1492034537",
      "title": "Smoke Cartel Crew Shirts",
    },
    Object {
      "availability": true,
      "compare_at": -1,
      "custom_fields": Object {},
      "hash": "AV11Vmq7CQkVMwqrFH0q",
      "id": "10246101192",
      "price": Array [
        60,
      ],
      "product_url": "https://smoke-cartel.myshopify.com/products/smoke-cartel-backpack",
      "reviews": Object {
        "average_score": 0,
        "total_reviews": 0,
      },
      "stickers": Object {
        "availability": true,
        "discount": false,
        "free-shipping": false,
        "in-stock": true,
        "out-of-stock": false,
      },
      "thumbnail_url": "https://cdn.shopify.com/s/files/1/0350/8253/products/SMOKE-CARTEL-BACKPACK_SC-BACKPACK-BK-5_small.jpg?v=1500673539",
      "title": "Smoke Cartel Odor-Resistiant Backpack",
    },
  ],
  "meta": Object {
    "item_limit": 3,
    "q": "smok",
    "rid": "<stripped>",
    "suggestion_limit": 5,
  },
  "suggestions": Array [
    Object {
      "value": "smoke",
    },
    Object {
      "value": "smoke cartel",
    },
    Object {
      "value": "smoky",
    },
    Object {
      "value": "smoky mountain",
    },
    Object {
      "value": "smoky mountain glass",
    },
  ],
}
`;

exports[`Client autocomplete respects request parameters 3`] = `
Object {
  "items": Array [
    Object {
      "availability": true,
      "compare_at": -1,
      "custom_fields": Object {},
      "hash": "AV11VmrFCQkVMwqrFH9w",
      "id": "4595632644",
      "price": Array [
        31,
      ],
      "product_url": "https://smoke-cartel.myshopify.com/products/5-rasta-frit-pipe",
      "reviews": Object {
        "average_score": 5,
        "total_reviews": 3,
      },
      "stickers": Object {
        "availability": true,
        "discount": false,
        "free-shipping": false,
        "in-stock": true,
        "out-of-stock": false,
      },
      "thumbnail_url": "https://cdn.shopify.com/s/files/1/0350/8253/products/HP62-1-0_small.jpg?v=1459019662",
      "title": "Rasta-fire-ian Pipe",
    },
    Object {
      "availability": true,
      "compare_at": 36,
      "custom_fields": Object {},
      "hash": "AV11VmrFCQkVMwqrFH9x",
      "id": "4595632836",
      "price": Array [
        27,
      ],
      "product_url": "https://smoke-cartel.myshopify.com/products/3-75-rasta-frit-pipe",
      "reviews": Object {
        "average_score": 4.83,
        "total_reviews": 6,
      },
      "stickers": Object {
        "availability": true,
        "discount": Array [
          25,
        ],
        "free-shipping": false,
        "in-stock": true,
        "out-of-stock": false,
      },
      "thumbnail_url": "https://cdn.shopify.com/s/files/1/0350/8253/products/HP61-1-0_small.jpg?v=1459021436",
      "title": "Rasta-fire-ian Jr.",
    },
  ],
  "meta": Object {
    "item_limit": 2,
    "q": "fir",
    "rid": "<stripped>",
    "suggestion_limit": 10,
  },
  "suggestions": Array [
    Object {
      "value": "fire",
    },
    Object {
      "value": "fire ian",
    },
    Object {
      "value": "fire cut",
    },
    Object {
      "value": "fire cut downstem",
    },
    Object {
      "value": "fire rasta",
    },
    Object {
      "value": "fire ian jr",
    },
    Object {
      "value": "fire 14.5mm",
    },
    Object {
      "value": "fire 18mm",
    },
    Object {
      "value": "fire 14.5mm 18.8mm",
    },
    Object {
      "value": "fire 18mm 18mm",
    },
  ],
}
`;
